File: app/src/main/java/ru/melodin/fast/current/BaseFragment.java
package ru.melodin.fast.current;

import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.AppBarLayout;

import ru.melodin.fast.view.FastToolbar;

public abstract class BaseFragment extends Fragment {

    private CharSequence title;
    private RecyclerView recyclerView;
    private FastToolbar toolbar;

    public FastToolbar getToolbar() {
        return toolbar;
    }

    public void setToolbar(FastToolbar toolbar) {
        this.toolbar = toolbar;
    }

    public RecyclerView getRecyclerView() {
        return recyclerView;
    }

    protected void setRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
    }

    protected CharSequence getTitle() {
        return title;
    }

    protected void setTitle(String title) {
        this.title = title;
    }

    public void scrollToTop() {
        if (recyclerView != null)
            recyclerView.smoothScrollToPosition(0);

        if (toolbar != null)
            expandToolbar();
    }

    private void expandToolbar() {
        AppBarLayout appBar = (AppBarLayout) toolbar.getParent();
        CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) appBar.getLayoutParams();
        AppBarLayout.Behavior behavior = (AppBarLayout.Behavior) params.getBehavior();
        if (behavior != null) {
            behavior.setTopAndBottomOffset(0);
            behavior.onNestedPreScroll((CoordinatorLayout) appBar.getParent(), appBar, null, 0, 1, new int[2]);
        }
    }
}

File: app/src/main/java/ru/melodin/fast/current/BaseFragment.kt
package ru.melodin.fast.current

import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.MaterialToolbar
import ru.melodin.fast.MainActivity

abstract class BaseFragment : Fragment() {

    internal var title: CharSequence? = null

    var recyclerList: RecyclerView? = null
    var toolbar: MaterialToolbar? = null

    var parent: MainActivity? = null

    open fun isBottomViewVisible(): Boolean {
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        parent = activity as MainActivity?
    }

    fun scrollToTop() {
        if (recyclerList != null)
            recyclerList!!.smoothScrollToPosition(0)

        if (toolbar != null)
            expandToolbar()
    }

    private fun expandToolbar() {
        val appBar = toolbar!!.parent as AppBarLayout
        val params = appBar.layoutParams as CoordinatorLayout.LayoutParams
        val behavior = params.behavior as AppBarLayout.Behavior?
        if (behavior != null) {
            behavior.topAndBottomOffset = 0
            behavior.onNestedPreScroll(
                appBar.parent as CoordinatorLayout,
                appBar,
                appBar,
                0,
                1,
                IntArray(2)
            )
        }
    }

    fun drawable(@DrawableRes resId: Int): Drawable {
        return ContextCompat.getDrawable(context!!, resId)!!
    }

    @ColorInt
    fun color(@ColorRes resId: Int): Int {
        return ContextCompat.getColor(context!!, resId)
    }

    fun string(@StringRes resId: Int): String {
        return getString(resId)
    }

    fun string(@StringRes resId: Int, vararg args: Any): String {
        return getString(resId, *args)
    }

    open fun onBackPressed() {}

}
